### PR TITLE
JoeMakerV4

### PR DESCRIPTION
**Pull request by LouisMeMyself**
Added slippage check before swapping to avoid low liquidity pair (mainly because of bad routing).

### CODE REVIEW

File: test/JoeMakerV4.test.ts [hunk 1]
```diff
@@ -115,7 +115,7 @@ describe("joeMakerV4", function () {
     it("should convert WAVAX", async function () {
       await this.wavax.deposit({ value: ethers.utils.parseEther("2") })
       await this.wavax.transfer(this.joeMakerV4.address, await this.wavaxERC20.balanceOf(this.dev.address))
-      await this.joeMakerV4.convert(this.wavax.address, this.wavax.address)
+      await this.joeMakerV4.convert(this.wavax.address, this.wavax.address, "100")
       expect(await this.usdc.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.wavaxERC20.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.usdc.balanceOf(BAR_ADDRESS)).to.equal("215434297")
```
Comment (cryptofish7) [inline test/JoeMakerV4.test.ts:121]: I don't quite get what you're testing here

File: test/JoeMakerV4.test.ts [hunk 2]
```diff
@@ -125,7 +125,7 @@ describe("joeMakerV4", function () {
       await this.router.swapExactAVAXForTokens("0", [this.wavax.address, this.usdc.address], this.joeMakerV4.address, DEADLINE, {
         value: ethers.utils.parseEther("2"),
       })
-      await this.joeMakerV4.convert(this.usdc.address, this.usdc.address)
+      await this.joeMakerV4.convert(this.usdc.address, this.usdc.address, "100")
       expect(await this.usdc.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.wavaxERC20.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.usdc.balanceOf(BAR_ADDRESS)).to.equal("215434297")
```
Comment (cryptofish7) [inline test/JoeMakerV4.test.ts:131]: Don't get what you're testing here. Shouldn't it test that USDC was converted to JOE and the balance of JOE held by Bar was increased by an expected amount?

File: test/JoeMakerV4.test.ts [hunk 3]
```diff
@@ -135,7 +135,7 @@ describe("joeMakerV4", function () {
       await this.router.swapExactAVAXForTokens("0", [this.wavax.address, this.wbtc.address], this.joeMakerV4.address, DEADLINE, {
         value: ethers.utils.parseEther("2"),
       })
-      await this.joeMakerV4.convert(this.wbtc.address, this.wbtc.address)
+      await this.joeMakerV4.convert(this.wbtc.address, this.wbtc.address, "100")
       expect(await this.usdc.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.wavaxERC20.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.usdc.balanceOf(BAR_ADDRESS)).to.equal("214143468")
```
Comment (cryptofish7) [inline test/JoeMakerV4.test.ts:141]: Same as above comment for USDC

File: test/JoeMakerV4.test.ts [hunk 4]
```diff
@@ -146,7 +146,7 @@ describe("joeMakerV4", function () {
     it("should convert AVAX - USDC", async function () {
       await this.zap.zapIn(this.avaxUsdc.address, { value: ethers.utils.parseEther("2") })
       await this.avaxUsdc.transfer(this.joeMakerV4.address, await this.avaxUsdc.balanceOf(this.dev.address))
-      await this.joeMakerV4.convert(this.usdc.address, this.wavax.address)
+      await this.joeMakerV4.convert(this.usdc.address, this.wavax.address, "100")
       expect(await this.usdc.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.avaxUsdc.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.usdc.balanceOf(BAR_ADDRESS)).to.equal("215111242")
```
Comment (cryptofish7) [inline test/JoeMakerV4.test.ts:152]: Again, I think we need better assertions here; the expectation is that Bar receives a certain amount of JOE

File: test/JoeMakerV4.test.ts [hunk 5]
```diff
@@ -155,7 +155,7 @@ describe("joeMakerV4", function () {
     it("should convert USDC - DAI", async function () {
       await this.zap.zapIn(this.usdcDai.address, { value: ethers.utils.parseEther("2") })
       await this.usdcDai.transfer(this.joeMakerV4.address, await this.usdcDai.balanceOf(this.dev.address))
-      await this.joeMakerV4.convert(this.dai.address, this.usdc.address)
+      await this.joeMakerV4.convert(this.dai.address, this.usdc.address, "100")
       expect(await this.usdc.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.usdcDai.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.usdc.balanceOf(BAR_ADDRESS)).to.equal("214878919")
```
Comment (cryptofish7) [inline test/JoeMakerV4.test.ts:161]: Again, I think we need better assertions here; the expectation is that Bar receives a certain amount of JOE

File: test/JoeMakerV4.test.ts [hunk 6]
```diff
@@ -164,7 +164,7 @@ describe("joeMakerV4", function () {
     it("should convert JOE - AVAX", async function () {
       await this.zap.zapIn(this.joeAvax.address, { value: ethers.utils.parseEther("2") })
       await this.joeAvax.transfer(this.joeMakerV4.address, await this.joeAvax.balanceOf(this.dev.address))
-      await this.joeMakerV4.convert(this.joe.address, this.wavax.address)
+      await this.joeMakerV4.convert(this.joe.address, this.wavax.address, "100")
       expect(await this.usdc.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.joeAvax.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.usdc.balanceOf(BAR_ADDRESS)).to.equal("214466109")
```
Comment (cryptofish7) [inline test/JoeMakerV4.test.ts:170]: Again, I think we need better assertions here; the expectation is that Bar receives a certain amount of JOE

File: test/JoeMakerV4.test.ts [hunk 7]
```diff
@@ -175,7 +175,7 @@ describe("joeMakerV4", function () {
       await this.mimTime.transfer(this.joeMakerV4.address, await this.mimTime.balanceOf(this.dev.address))
       await this.joeMakerV4.setBridge(this.time.address, this.mim.address)
       await this.joeMakerV4.setBridge(this.mim.address, this.wavax.address)
-      await this.joeMakerV4.convert(this.mim.address, this.time.address)
+      await this.joeMakerV4.convert(this.mim.address, this.time.address, "100")
       expect(await this.usdc.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.mimTime.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.usdc.balanceOf(BAR_ADDRESS)).to.equal("213737823")
```
Comment (cryptofish7) [inline test/JoeMakerV4.test.ts:181]: Again, I think we need better assertions here; the expectation is that Bar receives a certain amount of JOE

File: test/JoeMakerV4.test.ts [hunk 8]
```diff
@@ -206,41 +206,44 @@ describe("joeMakerV4", function () {
         value: ethers.utils.parseEther("1"),
       })
 
-      await this.joeMakerV4.convert(this.tractor.address, this.wavax.address)
+      await this.joeMakerV4.convert(this.tractor.address, this.wavax.address, "100")
 
       expect(await this.usdc.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.usdcDai.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.usdc.balanceOf(BAR_ADDRESS)).to.equal("186168809")
```
Comment (cryptofish7) [inline test/JoeMakerV4.test.ts:213]: Again, I think we need better assertions here; the expectation is that Bar receives a certain amount of JOE

File: test/JoeMakerV4.test.ts [hunk 9]
```diff
@@ -206,41 +206,44 @@ describe("joeMakerV4", function () {
...
-      const exploiter = await exploiterCF.deploy(this.joeMakerV4.address)
-      await exploiter.deployed()
+    it("reverts if using a pair with really low liquidity even with slippage at maximum", async function () {
+      await this.zap.zapIn(this.mimTime.address, { value: "2000000000000000000" })
+      await this.mimTime.transfer(this.joeMakerV4.address, await this.mimTime.balanceOf(this.dev.address))
+      await this.joeMakerV4.setBridge(this.mim.address, this.joe.address)
+      await expect(this.joeMakerV4.convert(this.mim.address, this.time.address, "4999")).to.be.revertedWith("JoeMakerV4: Slippage caught") // slippage at 1% this will revert as the JOE/MIM pair has really low liquidity.
+    })
 
+    it("reverts if convert is called by non auth", async function () {
```
Comment (cryptofish7) [inline test/JoeMakerV4.test.ts:229]: `reverts if convert is called by non-authorised user`

File: test/JoeMakerV4.test.ts [hunk 10]
```diff
@@ -250,7 +253,7 @@ describe("joeMakerV4", function () {
       await this.zap.zapIn(this.avaxUsdc.address, { value: ethers.utils.parseEther("2") })
       await this.joeAvax.transfer(this.joeMakerV4.address, await this.joeAvax.balanceOf(this.dev.address))
       await this.avaxUsdc.transfer(this.joeMakerV4.address, await this.avaxUsdc.balanceOf(this.dev.address))
-      await this.joeMakerV4.convertMultiple([this.joe.address, this.usdc.address], [this.wavax.address, this.wavax.address])
+      await this.joeMakerV4.convertMultiple([this.joe.address, this.usdc.address], [this.wavax.address, this.wavax.address], "100")
       expect(await this.usdc.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.joeAvax.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.avaxUsdc.balanceOf(this.joeMakerV4.address)).to.equal(0)
```
Comment (cryptofish7) [inline test/JoeMakerV4.test.ts:259]: Again, I think we need better assertions here; the expectation is that Bar receives a certain amount of JOE

File: test/JoeMakerV4.test.ts [hunk 11]
```diff
@@ -270,7 +273,7 @@ describe("joeMakerV4", function () {
 
       await this.zap.zapIn(this.avaxUsdc.address, { value: ethers.utils.parseEther("2") })
       await this.avaxUsdc.transfer(this.joeMakerV4.address, await this.avaxUsdc.balanceOf(this.dev.address))
-      await this.joeMakerV4.convert(this.usdc.address, this.wavax.address)
+      await this.joeMakerV4.convert(this.usdc.address, this.wavax.address, "100")
       expect(await this.usdc.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect(await this.avaxUsdc.balanceOf(this.joeMakerV4.address)).to.equal(0)
       expect((await this.usdc.balanceOf(BAR_ADDRESS)) - barBalance).to.be.greaterThan(0)
```
Comment (cryptofish7) [inline test/JoeMakerV4.test.ts:279]: Again, I think we need better assertions here; the expectation is that devs receive an expected amount of JOE

File: test/JoeMakerV4.test.ts [hunk 12]
```diff
@@ -293,7 +296,7 @@ describe("joeMakerV4", function () {
       expect(await this.joe.balanceOf(BAR_ADDRESS)).to.equal("90009160457674355334497135")
 
       await this.avaxUsdc.transfer(this.joeMakerV4.address, await this.avaxUsdc.balanceOf(this.dev.address))
-      await this.joeMakerV4.convert(this.usdc.address, this.wavax.address)
+      await this.joeMakerV4.convert(this.usdc.address, this.wavax.address, "100")
       expect(await this.joe.balanceOf(BAR_ADDRESS)).to.equal("90009210905950105791950559")
```
Comment (cryptofish7) [inline test/JoeMakerV4.test.ts:300]: What's the difference increase in JOE? Let's assert that it equals to that instead of a random number

**Review by cryptofish7 — CHANGES_REQUESTED** [review]

File: test/JoeMakerV4.test.ts (hunk 1 above)
Comment (LouisMeMyself) [reply test/JoeMakerV4.test.ts:121]: JoeMakerV4 uses USDC

File: contracts/JoeMakerV4.sol [hunk 13]
```diff
@@ -260,6 +297,14 @@ contract JoeMakerV4 is Ownable {
...
 
         amountOut = getAmountOut(amountInput, reserveInput, reserveOutput);
+
+        {
+            uint256 rest = uint256(10_000).sub(slippage);
+            require(
+                getAmountOut(amountOut, reserveOutput, reserveInput) > amountInput.mul(rest).mul(rest).div(100_000_000),
+                "JoeMakerV4: Slippage caught"
+            );
+        }
```
[the remark this answers is not in the document]
Comment (LouisMeMyself) [reply contracts/JoeMakerV4.sol:307]: (it should be getAmountOut(amountOut, reserveOutput, reserveInput) **>=** amountInput.mul(rest).mul(rest).div(100_000_000))

As we're simulating swapping in and out, slippage would happen twice.
if we tolerate 1% slippage for example and 1 AVAX = 100 USDC,
1 AVAX -> 99 USDC, this is 1% slippage, but as we calculate the swapping back to AVAX too,
99 USDC -> 98.01 USDC, that's 1.99% slippage with that method.
by doing `rest^2`, we're actually allowing 1.99% slippage on swap in / swap out, hence only 1% slippage on swap in

File: contracts/JoeMakerV4.sol [hunk 14]
```diff
@@ -260,24 +329,50 @@ contract JoeMakerV4 is Ownable {
         uint256 amountInput = IERC20(fromToken).balanceOf(address(pair)).sub(reserveInput); // calculate amount that was transferred, this accounts for transfer taxes
 
         amountOut = getAmountOut(amountInput, reserveInput, reserveOutput);
+
+        {
+            uint256 rest = uint256(10_000).sub(slippage);
+            /// @dev We calculate the amount received if we did a swapIn and swapOut, without updating the reserves,
+            /// this allows to catch if a pair has low liquidity. We do `rest^2`, i.e. calculating twice the slippage
+            /// cause we actually do two swaps, so slippage is actually applied twice.
```
Comment (cryptofish7) [inline contracts/JoeMakerV4.sol:337]:
```
/// @dev We simulate the amount received if we did a swapIn and swapOut without updating the reserves,
/// hence why we do rest^2, i.e. calculating the slippage twice cause we actually do two swaps.
/// This allows us to catch if a pair has low liquidity
```

File: contracts/JoeMakerV4.sol [hunk 15]
```diff
@@ -260,24 +329,50 @@ contract JoeMakerV4 is Ownable {
...
             ? (uint256(0), amountOut)
             : (amountOut, uint256(0));
         pair.swap(amount0Out, amount1Out, to, new bytes(0));
     }
 
-    // F1 - F10: OK
-    // C1 - C24: OK
-    function _toToken(address token, uint256 amountIn) internal returns (uint256 amountOut) {
+    /// @notice Swaps `amountIn` `token` to `tokenTo`, `amountOut` is required to be greater
+    /// than allowed `slippage`.
```
Comment (cryptofish7) [inline contracts/JoeMakerV4.sol:352]: I'd avoid using too many variables in the comment; it's important to explain the function in simple english.

```
/// @notice Swaps an amount of token to another token, `tokenTo`
/// @dev `amountOut` is required to be greater after slippage 
```

File: contracts/JoeMakerV4.sol [hunk 16]
```diff
@@ -260,24 +329,50 @@ contract JoeMakerV4 is Ownable {
...
+    /// @param token The address of token that will be swapped
+    /// @param amountIn The amount of the `token`
+    /// @param slippage The accepted slippage, In basis points aka parts per 10,000 so 5000 is 50%
```
Comment (cryptofish7) [inline contracts/JoeMakerV4.sol:355]: `In`->`in`

File: contracts/JoeMakerV4.sol [hunk 17]
```diff
@@ -260,24 +329,50 @@ contract JoeMakerV4 is Ownable {
...
+    /// @return amountOut The amount of `toToken` sent to `to`
```
Comment (cryptofish7) [inline contracts/JoeMakerV4.sol:356]: `sent to JoeBar`

File: contracts/JoeMakerV4.sol [hunk 18]
```diff
@@ -260,24 +329,50 @@ contract JoeMakerV4 is Ownable {
...
         if (devCut > 0) {
             amount = amount.mul(devCut).div(10000);
             IERC20(token).safeTransfer(devAddr, amount);
             amount = amountIn.sub(amount);
         }
-        amountOut = _swap(token, tokenTo, amount, bar);
+        amountOut = _swap(token, tokenTo, amount, bar, slippage);
     }
 
+    /// @notice Returns the amount that user will receive after swaping from `in` to `out`
```
Comment (cryptofish7) [inline contracts/JoeMakerV4.sol:371]: `swaping`->`swapping`

**Review by cryptofish7 — CHANGES_REQUESTED** [review]

**Review by cryptofish7 — APPROVED** [review]

**Pull request merged**